### PR TITLE
Adding some comments documenting captp.js.

### PR DESCRIPTION
**Pull request by cwebber**
First pass: there are some unanswered questions in the comments and some things are probably wrong!

I'd love review about how much I understand this... and I figured, while I'm trying to figure it out, maybe my attempt-at-understanding is generally useful for the module.

There are some questions (marked with `???`) amongst the comments and some things are probably flat-out wrong, so review is encouraged before merging.

I'm happy to make some changes if you are willing to review... or if you'd prefer to just modify and pull in yourself, that's welcome too.

### CODE REVIEW

**Comment by cwebber** [comment]
Also I don't know anything about Agoric's style guide (or contemporary Javascript style at all) so sorry for the failing linting tests :)

**Comment by cwebber** [comment]
BTW, an additional thing to possibly comment, once understood:

```
<Zarutian_HTC> dustyweb: one thing I noticed about CTP_RETURN that the question gets deleted. No idea why.
<Zarutian_HTC> seem to prevent usage patterns like: const t1 = remoteObj~.x(); t1.y(); t1.z();
<dustyweb> Zarutian_HTC: yes that is curious
<dustyweb> Zarutian_HTC: mind if I put that on the issue tracker?
<Zarutian_HTC> sure
<dustyweb> Zarutian_HTC: I have a guess
<dustyweb> it's an attempt to prevent things from getting too out of hand before GC support in captp
<dustyweb> .js
<Zarutian_HTC> right
<Zarutian_HTC> the gc thing requires weakref and post mortem finalization
```

Are we right? :)

**Comment by michaelfig** [comment]
> Are we right? :)

Probably! :)

BTW, your comments are welcome.  You can look at the current CapTP as me trying to read and understand the Cap'n'Proto implementation and the liveSlots.js implementation, but not as well-documented.

File: packages/captp/lib/captp.js [hunk 1]
```diff
@@ -10,31 +10,51 @@ export { E, HandledPromise, Nat, harden };
...
+  // Since we decide the numbers for questions, we use this to increment
+  // the question key
   let lastQuestionID = 0;
 
-  const valToSlot = new WeakMap();
-  const slotToVal = new Map(); // exports
-  const questions = new Map(); // chosen by us
-  const answers = new Map(); // chosen by our peer
-  const imports = new Map(); // chosen by our peer
+  const valToSlot = new WeakMap();   // exports looked up by val
```
Comment (michaelfig) [inline packages/captp/lib/captp.js]: These lint complaints are due to prettier only allowing a single space between the end of line and `//`.  Could you please fix this?

File: packages/captp/lib/captp.js [hunk 2]
```diff
@@ -103,8 +138,13 @@ export function makeCapTP(ourId, send, bootstrapObj = undefined) {
     return harden(pr);
   }
 
+  // Set up import
   function convertSlotToVal(theirSlot) {
     let val;
+    // Invert slot direction from other side.
+    // (??? cwebber): Why is this inverted this way?  In what circumstance
```
Comment (michaelfig) [inline packages/captp/lib/captp.js]: The direction prevents namespace collisions between the slots we allocate and the ones that the other side allocates.  If we allocate a slot, serialise it to the other side, and they send it back to us, we need to reference just our own slot, not one from their side.

File: packages/captp/lib/captp.js [hunk 3]
```diff
@@ -138,7 +180,14 @@ export function makeCapTP(ourId, send, bootstrapObj = undefined) {
         result: serialize(bootstrap),
       });
     },
+    // Remote is invoking a method or retrieving a property.
     CTP_CALL(obj) {
+      // questionId: Remote promise (for promise pipelining) this call is
+      //   to fulfill
+      // target: Slot id of the target to be invoked.  Checks against
+      //   answers first; otherwise goes through unserializer
+      // (??? cwebber): Not sure that's well phrased or even if I understand
```
Comment (michaelfig) [inline packages/captp/lib/captp.js]: Yup, you understand it at least as well as it was written.

**Review by michaelfig — CHANGES_REQUESTED** [review]
Just a few changes.  Thanks for this!

**Comment by cwebber** [comment]
Updated.  Thank you!

(I don't completely understand *why* the directionality is introduced however... my read on the imports/exports and questions/answers per reading the original captp documentation is that the imports/exports end up being "specified" by the exporter and that questions/answers are "specified" by the questioner, and this is relevant to the way that collaborative GC ends up happening.  But maybe I've misunderstood something... anyway, not critical to getting this PR in.)

**Review by michaelfig — APPROVED** [review]

**Pull request merged**